doc/rfc008-managing-apalache.md

```diff
@@ -0,0 +1,267 @@
+
+- Invoking the system-level package manager from Quint (see §3.2.1)
+- Invoking an ecosystem package manager from Quint (see §3.2.2)
+- Pull the Apalache distribution from Github (see §3.2.3), e.g. using
+  [Github's REST API endpoint][] and/or [`octokit/request.js`][]
+
+Pros:
+
+- Completely transparent to the user.
+- Allows managing the required Apalache version from within Quint.
+
+Cons:
+
+- Only pushes the boundary: e.g., this would assume that npm, coursier, or the
+  JRE are installed on the local system. `vscode-metals` includes a
+  `coursier` binary in its distribution to counter this.
+- Spawning a package manager increases complexity by running a process whose
+  behavior is outside of our control.
+
+#### 3.2.5 Apalache as cloud-hosted SaaS
+
+Quint uses Apalache via a public SaaS API, no local installation or management
+is necessary.
+
+Pros:
+
+- No installation or server management issues.
+- Versioning and backwards-compatiblity could be exposed via separate endpoints.
+
+Cons:
+
+- Relies on internet connection.
+- apalache-cloud is currently not exposed via the Shai gRPC interface.
+- Recurring cost for cloud hosting, likely requires billing of external users.
+- May deter external users for privacy concerns.
+
+### 3.3 Version dependencies between Quint and Apalache
+
+How do we manage version depenedencies between Quint and Apalache?
+
+_TBD: this depends heavily on the solution to §3.1 and §3.2, and should be
+determined after we reach consensus on those points._
+
+## 4. Proposed Solution
+
+We choose managed options for obtaining and launching Apalache from Quint, while
+still providing a non-managed option:
+
+1. Primarily, Quint uses a Quint-managed, on-demand instance of Shai:
+   1. If not present, Quint fetches a compatible Apalache release from Github,
+      using [Github's REST API endpoint][] and/or [`octokit/request.js`][] and
+      unpacks it into a local installation directory.
+      - This can be scripted etirely from NodeJS, foregoing the need to wrap a
+        third-party package manager binary.
+   2. Apalache's JRE dependency is taken care of by adding a check for a `java`
+      executable in `$PATH` to [the `apalache-mc` runner script](https://github.com/informalsystems/apalache/blob/df7adb8b42b6487de9764162f338935121d07a3c/src/universal/bin/apalache-mc#L53).
+      - This shall print instructions for obtaining a JRE, if none is detected.
+2. Quint launches an on-demand instance of this local installation by spawning
+   `apalache-mc server` in a separate process.
+   - Such a process is spawned for each invocation of the `verify` command, and
+     torn down immediately after the RPC response.
+     - This seems to be a reasonable tradeoff given Apalache's memory
+       consumption and JVM startup time.
+     - We can later change this to a long-lived server, if we use a more
+       stateful approach of interacting with Shai (e.g., via the [transition
+       explorer API][]).
```

Comment:
In addition to this, we could have a special management inside the language server that keeps the apalache server running. Since not all language server users will need apalache, we could start apalache on the first request of a `verify`, and then keep it running until the language server is closed.

I don't think this is super necessary, surely not for now. But I wanted to point out that there is this language server scenario where I believe it makes sense to have the apalache server running continuously.

Reply:
That's great input! I'm still a bit weary of keeping a long-running Apalache server running, because most people will only use it intermittently.
But launching it from LSP is a good pointer! Definitely the preferred option once we're able to run Quint from VSCode.